core/peerset/src/lib.rs

```diff
@@ -236,68 +351,85 @@ impl PeersetMut {
 	///
 	/// Because of concurrency issues, it is acceptable to call `incoming` with a `PeerId` the
 	/// peerset is already connected to, in which case it must not answer.
-	pub fn incoming(&self, peer_id: PeerId, index: IncomingIndex) {
-		let mut inner = self.parent.inner.lock();
-		if inner.out_slots.iter().chain(inner.in_slots.iter()).any(|s| s.as_ref() == Some(&peer_id)) {
-			return
+	pub fn incoming(&mut self, peer_id: PeerId, index: IncomingIndex) {
+		// if `reserved_only` is set, but this peer is not a part of our discovered list,
+		// a) it is not reserved, so we reject the connection
+		// b) we are already connected to it, so we reject the connection
+		if self.data.reserved_only && !self.data.discovered.is_reserved(&peer_id) {
+			self.message_queue.push_back(Message::Reject(index));
+			return;
+		}
+
+		// check if we are already connected to this peer
+		if self.data.out_slots.contains(&peer_id) {
+			self.message_queue.push_back(Message::Reject(index));
+			return;
 		}
 
-		if let Some(pos) = inner.in_slots.iter().position(|s| s.is_none()) {
-			inner.in_slots[pos] = Some(peer_id);
-			let _ = self.parent.tx.unbounded_send(Message::Accept(index));
+		let slot_type = if self.data.reserved_only {
+			SlotType::Reserved
 		} else {
-			if inner.discovered.iter().all(|p| *p != peer_id) {
-				inner.discovered.push(peer_id);
+			SlotType::Common
+		};
+
+		loop {
+			match self.data.in_slots.add_peer(peer_id.clone(), slot_type) {
+				Ok(_) => {
+					self.data.discovered.remove_peer(&peer_id);
+					self.message_queue.push_back(Message::Accept(index));
+					return;
+				},
+				Err(SlotError::MaxConnections(peer_id)) => {
+					self.data.discovered.add_peer(peer_id, slot_type);
+					self.message_queue.push_back(Message::Reject(index));
+					return;
+				},
+				Err(SlotError::AlreadyConnected(_)) => {
+					self.message_queue.push_back(Message::Reject(index));
+					return;
+				},
+				Err(SlotError::DemandReroute { disconnect, .. }) => {
+					// disconnect not reserved node
```

Comment:
Should send back `Accept` here, no?

Reply:
no, we are in the loop. `Accept` will be send on the next loop iteration